### ISSUE TITLE
Document how to decrypt and extract files from backups

### ISSUE THREAD

**Issue by SebiderSushi** [stevesoltys/seedvault#104]
In some cases it can be necessary to access information from a backup, but restoring the backup to a device and using its corresponding app to access the information is not viable.

For example, a specific information might only be stored inside a backup because it has never been required outside of the context of the application or added just recently. At some point, the device storing the app & original copy of the information breaks and the data becomes inaccessible. Repairing the device or getting a new one to restore the backup and access the information again is either not possible/desired or the information is urgently needed and has to be acquired as soon as possible.
&nbsp;In such a case, a knowledgeable user could use a PC or other capable device to decrypt an existing backup and extract relevant files to manually recover the required piece of information.

**Comment by grote** [comment]
Documentation won't help you much. You need a decrypter tool which could probably be built from seedvault's kotlin sources. However, I would consider such a tool a separate project.

**Comment by grote** [comment]
Also before someone undertakes this separate project: Consider the value of being able to decrypt backups off-device. What is encrypted are essentially *binary blobs* that the system extracts or gets from the apps. You will need further specialized tools to extract something meaningful from them and once you have that, you have an app specific data structure that again might require another specialized tool to get to the information you need.

**Comment by SebiderSushi** [comment]
Wait this backup utility has absolutely no access to plain files from the application directories then?

**Comment by grote** [comment]
It interfaces with the AOSP backup API which essentially sends a byte stream for us to store.

**Comment by grote** [comment]
@SebiderSushi for that reason and the separate project one, I am inclined to close this ticket. WDYT?

**Comment by chirayudesai** [comment]
I feel like this could still be nice to have, just to verify the backup even.

Since we do use the same basic format as 'adb backup', there could already be some existing tool to parse the decrypted data.

We are also talking about backing up external storage, and if that's encrypted by the same mechanism decryption of that could be useful too.

One could take the existing code here and put it into a small java/kotlin program, that way might be the fastest.

Alternatively, if somebody does an alternate implementation, that could be good to have in general as well just to verify our crypto code (and catch any implementation / language bugs for example)

I'd be inclined to just keep this open.

**Comment by SebiderSushi** [comment]
> @SebiderSushi for that reason and the separate project one, I am inclined to close this ticket. WDYT?

@grote 
If there is no file access possible then i wouldn't know what to make of it anyways :/
But if there are ways to extract or decompile these backups like @chirayudesai hinted then such a project would be great.

Anyhow i'd leave it up to you to decide whether this should better be tracked elsewhere.

**Comment by grote** [comment]
It looks like full backup data is just a tar file with the directory structure of the app's data dirs that were included in backup. I haven't verified this with a PoC, just got this impression from reading code.

**Comment by tlambertz** [comment]
I have written a simple python script to decrypt the SeedVault backups. You can find it at [seedvault_backup_parser](https://github.com/tlambertz/seedvault_backup_parser). It currently does not decompress the tar archives of full app backups.

I have mainly used it to modify the KV backups to import lots of wifi passwords.

**Comment by grote** [comment]
Awesome @tlambertz, great work! 

So the only thing left here is to add a link to that project in the README.

**Comment by norpol** [comment]
For others who end up on this page (since it's highest ranked on google).

The python tool `seedvault_backup_parser` does not support the current  `version 1` format.

https://github.com/tlambertz/seedvault_backup_parser/issues/14#issue-1147327900

There is the golang based project https://github.com/jackwilsdon/seedvault-extractor which unfortunately misses the "storage" type, but works for the apps that are being backed up.

**Comment by crass** [comment]
> The python tool `seedvault_backup_parser` does not support the current `version 1` format.
> 
> [tlambertz/seedvault_backup_parser#14 (comment)](https://github.com/tlambertz/seedvault_backup_parser/issues/14#issue-1147327900)
> 
> There is the golang based project https://github.com/jackwilsdon/seedvault-extractor which unfortunately misses the "storage" type, but works for the apps that are being backed up.

I've updated seedvault_backup_parser (see [this PR](https://github.com/tlambertz/seedvault_backup_parser/pull/16)) to now support decrypting V1 backups. It also supports copying APKs and decrypting the storage files. Encryption shouldn't be hard to add.

**Comment by grote** [comment]
Relevant here: For some research I wrote a small class that is able to plaintext save an existing backup to normal files on the phone storage: https://github.com/grote/seedvault/commit/95f062ee47e71ef78d8b5fdb799da0845e109a78#diff-b4ee7d18e13c7b9851f48b655459dd8b1f815561b92eb414a5519ee004faa7ccR25

It depends on other classes in Seedvault, but all could be extracted into a standalone utility. In the future I might move related classes into a seedvault-core module which would make this easier without having to vendorize Seedvault classes.

**Comment by khimaros** [comment]
@grote any tips on how to build this and expose to the user? do you think it would be feasible to target non-android platforms?

**Comment by grote** [comment]
With Kotlin multi-platform, you should be able to target other platforms like native Linux. However, the code indirectly depends on JVM classes like `KeyStore` for decryption. However, you could make a `.jar` that can run on Linux, Windows and Mac as well, if the JVM is installed. Then you'd just start with the code I linked above and step by step add all the other classes. But not sure what is faster, doing a python program from scratch or this.

**Comment by pwall2222** [comment]
So after scouring the internet, I have come to find that there is no comfortable way to decrypt v2 backups, so far. This should probably be mentioned on the README along side the tools to decrypt v0 and v1, because having to search through a myriad of issues is unpleasant.
And for anyone that found this comment, the only aiding help I can give is the fact that you could use an Android Emulator to get at least your files back.

**Comment by NicholasFlamy** [comment]
I found this because Seedvault isn't restoring files or apks from my backups. Now, I'm not sure what to do.

**Comment by LeonEuler** [comment]
@NicholasFlamy The seedvault-app readme was recently updated to have a link to Seednaut, a tool capable of inspecting and extracting v2 backups, including files.

Worked flawlessly a few days ago to help me confirm an app's backup had the json settings I hoped were included.

(I went the docker route with the official `rust` image, opened a bash inside the container, and `cargo install seednaut` installed the utility without any difficulty.)

Seednaut might not necessarily solve your problem, but it might be able to help confirm whether or not the files are there in the backup.

**Comment by NicholasFlamy** [comment]
> [@NicholasFlamy](https://github.com/NicholasFlamy) The seedvault-app readme was recently updated to have a link to Seednaut, a tool capable of inspecting and extracting v2 backups, including files.
> 
> Worked flawlessly a few days ago to help me confirm an app's backup had the json settings I hoped were included.
> 
> (I went the docker route with the official `rust` image, opened a bash inside the container, and `cargo install seednaut` installed the utility without any difficulty.)
> 
> Seednaut might not necessarily solve your problem, but it might be able to help confirm whether or not the files are there in the backup.

Thanks for the info. Also, I have a phone running LineageOS that was backed up and a phone running GrapheneOS that was backed up. A fresh GrapheneOS install is not able to install the apks from the LineageOS backup but IS able to install from the GrapheneOS backup.

It is NOT able to restore files from either backup during initial setup. It finds the file backups from LineageOS only after setup and clicking 3 dots -> restore -> "Skip restoring apps"

Edit: this is after enabling backing up files, without enabling it doesn't show up in restore

**Comment by grote** [comment]
there was a similar issue with that ROM for APK installs. Maybe they broke file backups as well.

**Comment by NicholasFlamy** [comment]
> there was a similar issue with that ROM for APK installs. Maybe they broke file backups as well.

Huh, I wonder what's up with that.

Also, to be clear, the issue I noted with file backups is UI/UX relating to seedvault where it wants to do apps first and it is supposed to show you files after restoring apps but doesn't during initial setup.

**Comment by grote** [comment]
>  the issue I noted with file backups is UI/UX relating to seedvault where it wants to do apps first and it is supposed to show you files after restoring apps but doesn't during initial setup.

not sure I understand the issue then. Could you please open a new ticket for that with a screenrecording showing the problem?

**Comment by NicholasFlamy** [comment]
> not sure I understand the issue then. Could you please open a new ticket for that with a screenrecording showing the problem?

The problem is, I don't have a device I can factory reset and clean install to.

**Comment by grote** [comment]
So you are saying that under some circumstances during SUW restore Seedvault is not offering to restore files, even if files backup is available to restore? Would be nice to know more about these circumstances so we could fix this issue.

**Comment by NicholasFlamy** [comment]
> So you are saying that under some circumstances during SUW restore Seedvault is not offering to restore files, even if files backup is available to restore?

Yes.

> Would be nice to know more about these circumstances so we could fix this issue.

I may look into running GrapheneOS in an emulator. I doubt it'll work considering all of the Pixel hardware specifics and security features though.
